### DISCUSSION TITLE
static x-ref on dynamically created element not working

### DISCUSSION THREAD

**Discussion by mokhosh** [alpinejs/alpine#2332] [7. Bugs (archived)] [answered]
I'm using filepond, and I have a custom "labelIdle" that has an x-ref. Something like this:
```html
<div
    wire:ignore
    x-data="{ pond : null }"
    x-init="
        pond = FilePond.create($refs.input, {
            labelIdle: `
                <span x-ref='label'>Drag and Drop or Select a Jpeg or Png Image</span>
            `,
        })
    "
    x-on:image-changed.window="
        if (! $event.detail.src)
        $refs.label.text = $event.detail.src
    "
>
```

But $refs doesn't have `label` in it. Is this because of the new thing in v3 where you can't bind x-ref?
https://alpinejs.dev/upgrade-guide#x-ref-no-more-dynamic

**Comment by SimoTod** [comment, marked as answer]
Where is your x-ref="label"? Also refs are dom element, I don't think they have a text property

**Reply by mokhosh** [reply]
x-ref label is defined in the element that i pass to filepond on initialisation.

Yeah, I've changed the example to make it simpler, elements not having text property is not the issue 😁 it's an img and I'm changing its src in the actual code. It worked perfectly on v2.

**Reply by SimoTod** [reply]
Right, so it gets injected insode the component. Try to delay it a bit (NextTick or setTimeout), i haven't tried but it could be that because the element hasn't fully initialised, it won't pick up new directive injected by x-init.

**Reply by mokhosh** [reply]
I tried $nextTick but that didn't work. I'll try timeout next and update you.
So you're saying this is not technically dynamic x-ref that is expected to not work. right?

**Reply by SimoTod** [reply]
Yeah, for dynamic Caleb meant something like `x-bind:x-ref="whatever" which use to work in v2 but it doesn't in v3. New dom element added dynamically should be picked up by the mutation observer and evaluates directives as usual. I'll try tonight and let you know

**Reply by SimoTod** [reply]
Yeah, for some reason, it doesn't work with FilePond even if it should work (this vanilla js example does about the same thing. https://codepen.io/SimoTod/pen/mdMKVNx). I couldn't figure out why though. You probably have to use a normal `document.querySelector()`.

**Reply by mokhosh** [reply]
Yeah that's what i did at the moment.
Worth digging in more?

**Reply by SimoTod** [reply]
Yeah, if you have time, feel free to have a look. I suspect it's something to do with FilePond moving the dom element and Alpine's mutation observer tearing down the directive. It's probably the same issue that causes #2329.

**Reply by mokhosh** [reply]
I doubt it's the same thing. Because if you downgrade to 3.0.8 then https://github.com/alpinejs/alpine/discussions/2329 works, but this still doesn't.

**Reply by mokhosh** [reply]
In either case I would say that change in 3.0.9 should be considered a big breaking change when it comes to integrating with libraries that manipulate dom elements, and should probably be reverted or something. No?

**Reply by mokhosh** [reply]
@SimoTod now that you have the answer to the other issue can you confirm if this is caused by the same issue?

**Reply by mokhosh** [reply]
Gonna mark this answered and continue the discussion here https://github.com/alpinejs/alpine/discussions/2329

**Reply by SimoTod** [reply]
It seems they are undesired effects of the same logic issue (although they are slightly different bugs). 
P.s. To clarify, 3.0.9 shouldn't be reverted. It fixed an issue that was causing memory leaks. By chance, the other bug was working before 3.0.9 because it was inadvertently exploiting one of those memory leaks but 3.0.9 is not the real root cause of those bugs.
